contracts/DutchAuction.sol

```diff
@@ -0,0 +1,176 @@
+                // Derive the duration for the order and place it on the stack.
+                duration = config.endTime - config.startTime;
+
+                // Derive time elapsed since the order started & place on stack.
+                elapsed = block.timestamp - config.startTime;
+
+                // Derive time remaining until order expires and place on stack.
+                remaining = duration - elapsed;
+            }
+
+            // Aggregate new amounts weighted by time with rounding factor.
+            // TODO: check the math boundary here
+            uint256 totalBeforeDivision = ((config.startAmountInWei *
+                remaining) + (config.endAmountInWei * elapsed));
+
+            // Use assembly to combine operations and skip divide-by-zero check.
+            assembly {
+                // Multiply by iszero(iszero(totalBeforeDivision)) to ensure
+                // amount is set to zero if totalBeforeDivision is zero,
+                // as intermediate overflow can occur if it is zero.
+                amount := mul(
+                    iszero(iszero(totalBeforeDivision)),
+                    // Subtract 1 from the numerator and add 1 to the result if
+                    // roundUp is true to get the proper rounding direction.
+                    // Division is performed with no zero check as duration
+                    // cannot be zero as long as startTime < endTime.
+                    add(
+                        div(sub(totalBeforeDivision, roundUp), duration),
+                        roundUp
+                    )
+                )
+            }
+
+            // Return the current amount.
+            return amount;
+        }
+
+        // Return the original amount as startAmount == endAmount.
+        return config.endAmountInWei;
+    }
+
+    function bid(uint32 qty)
+        external
+        payable
+        nonReentrant
+        hasSupply(qty)
+        validTime
+    {
+        uint256 price = getCurrentPriceInWei();
+        if (msg.value < qty * price) revert NotEnoughValue();
+
+        if (_refundable) {
+            User storage bidder = _userData[msg.sender]; // get user's current bid total
+            bidder.contribution = bidder.contribution + uint216(msg.value);
+            bidder.tokensBidded = bidder.tokensBidded + qty;
+
+            // _settledPriceInWei is always the minimum price of all the bids' unit price
+            if (price < _settledPriceInWei || _settledPriceInWei == 0) {
+                _settledPriceInWei = price;
+            }
+        }
+        _safeMint(msg.sender, qty);
+        emit Bid(msg.sender, qty, price);
+    }
+
+    function claimRefund() external nonReentrant {
+        Config memory config = _config;
+        if (!_refundable) revert NotRefundable();
+        if (config.endTime > block.timestamp) revert NotEnded();
+
+        User storage user = _userData[msg.sender];
+        if (user.refundClaimed) revert UserAlreadyClaimed();
```

Comment:
What if user

bid -> claimRefund -> bid -> claimRefund?

The second claimRefund would fail

Reply:
Proposed solution is to flip `user.refundClaimed` to true in `bid()`.

Reply:
user shouldn't be able to bid again after claiming refund, since you can only bid before `endTime` and you can only `claimRefund` after `endTime`

Reply:
Oh right, this is a false alarm.

Reply:
yeah, the entTime gates that

Reply:
I'm actually thinking if we just adopt this startTime/endTime for bucket auction as well, so it's consistent and we don't need to set additional 'isActive' stage, so to minimize the operation overhead